diff --git a/vmdebootstrap b/vmdebootstrap
index 126475f..a32c8f4 100755
--- a/vmdebootstrap
+++ b/vmdebootstrap
@@ -242,10 +242,12 @@ class VmDebootstrap(cliapp.Application):
         self.message('Creating partitions')
         self.runcmd(['parted', '-s', self.settings['image'],
                      'mklabel', 'msdos'])
+        # Start at 0% instead of 0 to allow parted to leave 1MiB for
+        # GRUB
         if self.settings['bootsize'] and self.settings['bootsize'] is not '0%':
             bootsize = str(self.settings['bootsize'] / (1024 * 1024))
             self.runcmd(['parted', '-s', self.settings['image'],
-                         'mkpart', 'primary', 'fat16', '0', bootsize])
+                         'mkpart', 'primary', 'fat16', '0%', bootsize])
         else:
             bootsize = '0%'
         self.runcmd(['parted', '-s', self.settings['image'],
@@ -486,6 +487,7 @@ class VmDebootstrap(cliapp.Application):
             self.runcmd(['chroot', rootdir, 'update-grub'])
             self.runcmd(['chroot', rootdir, 'grub-install', install_dev])
         except cliapp.AppException as e:
+            self.message(str(e))
             self.message("Failed. Is grub2-common installed? Using extlinux.")
         self.runcmd(['umount', os.path.join(rootdir, 'sys')])
         self.runcmd(['umount', os.path.join(rootdir, 'proc')])
